**Machine Learning Project**

Sum up of the project

Can we predict the risk of developing a tumor based on an individual’s symptoms? We aim to design a simple artificial intelligence system that determines whether a patient should be concerned given their observed symptoms. Using Python and basic machine learning tools, we will train a supervised classification model on a disease–symptom dataset.

*The data*

We use Disease-Symptom Dataset under World Bank license and found in https://www.kaggle.com/datasets/dhivyeshrk/diseases-and-symptoms-dataset. This dataset could not be used in a real research project, but is relevant for a project aimed at manipulating and understanding machine learning. It contains a list of 773 unique diseases associated with 377 symptoms. Theses symptoms are coded as dummy variable: 1 if the symptom is relevant to the disease, 0 otherwise. We end up with 246,000 samples.

**Taking a look at the data**

In [2]:
import pandas as pd

df = pd.read_csv("Final_Augmented_dataset_Diseases_and_Symptoms.csv")

# Je crée un sample pour que ça tourne plus vite
df_sample = df.sample(1000).copy()
df.sample(30)

diseases  anxiety and nervousness  \
4729                          eating disorder                        0   
225728                            skin cancer                        0   
59608                                sciatica                        0   
148994                        ear drum damage                        0   
118432                         fluid overload                        0   
222085                   personality disorder                        1   
209492                               neurosis                        0   
30802                             indigestion                        0   
226535                         aphthous ulcer                        0   
246594            conjunctivitis due to virus                        0   
65990                       injury to the leg                        0   
25780                            hyperkalemia                        0   
175308                       hormone disorder                        0   
238740  benign vaginal discharge (leukorrhea)                        0   
38240              deep vein thrombosis (dvt)                        0   
224083                           strep throat                        0   
131515         otitis externa (swimmer's ear)                        0   
166640                               migraine                        0   
34064                      acute otitis media                        0   
140497     benign prostatic hyperplasia (bph)                        0   
119495                          skin disorder                        0   
33731                       actinic keratosis                        0   
80100                              bell palsy                        0   
4699                          eating disorder                        1   
215231                     multiple sclerosis                        0   
117271                          ganglion cyst                        0   
215315                     multiple sclerosis                        0   
74832                   seborrheic dermatitis                        0   
180414                               cystitis                        0   
147029                                  croup                        0   

        depression  shortness of breath  depressive or psychotic symptoms  \
4729             1                    0                                 1   
225728           0                    0                                 0   
59608            0                    0                                 0   
148994           0                    0                                 0   
118432           0                    0                                 0   
222085           0                    0                                 1   
209492           0                    0                                 0   
30802            0                    0                                 0   
226535           0                    0                                 0   
246594           0                    0                                 0   
65990            0                    0                                 0   
25780            0                    1                                 1   
175308           0                    0                                 0   
238740           0                    0                                 0   
38240            0                    0                                 0   
224083           0                    0                                 0   
131515           0                    0                                 0   
166640           0                    0                                 0   
34064            0                    0                                 0   
140497           0                    0                                 0   
119495           0                    0                                 0   
33731            0                    0         

In [3]:
df.shape

(246945, 378)

There are 246,945 observations for 378 variables. We need to clean this a bit to make it more useful and relevant.

In [4]:
df.dtypes

diseases                            object
anxiety and nervousness              int64
depression                           int64
shortness of breath                  int64
depressive or psychotic symptoms     int64
                                     ...  
hip weakness                         int64
back swelling                        int64
ankle stiffness or tightness         int64
ankle weakness                       int64
neck weakness                        int64
Length: 378, dtype: object

In [5]:
df.describe()

anxiety and nervousness     depression  shortness of breath  \
count            246945.000000  246945.000000        246945.000000   
mean                  0.039235       0.042746             0.086440   
std                   0.194155       0.202285             0.281014   
min                   0.000000       0.000000             0.000000   
25%                   0.000000       0.000000             0.000000   
50%                   0.000000       0.000000             0.000000   
75%                   0.000000       0.000000             0.000000   
max                   1.000000       1.000000             1.000000   

       depressive or psychotic symptoms  sharp chest pain      dizziness  \
count                     246945.000000     246945.000000  246945.000000   
mean                           0.061001          0.097252       0.069943   
std                            0.239333          0.296302       0.255051   
min                            0.000000          0.000000       0.000000   
25%                            0.000000          0.000000       0.000000   
50%                            0.000000          0.000000       0.000000   
75%                            0.000000          0.000000       0.000000   
max                            1.000000          1.000000       1.000000   

            insomnia  abnormal involuntary movements  chest tightness  \
count  246945.000000                   246945.000000    246945.000000   
mean        0.039410                        0.040572         0.037871   
std         0.194568                        0.197296         0.190884   
min         0.000000                        0.000000         0.000000   
25%         0.000000                        0.000000         0.000000   
50%         0.000000                        0.000000         0.000000   
75%         0.000000                        0.000000         0.000000   
max         1.000000                        1.000000         1.000000   

        palpitations  ...  stuttering or stammering  problems with orgasm  \
count  246945.000000  ...                  246945.0              246945.0   
mean        0.024876  ...                       0.0                   0.0   
std         0.155747  ...                       0.0                   0.0   
min         0.000000  ...                       0.0                   0.0   
25%         0.000000  ...                       0.0                   0.0   
50%         0.000000  ...                       0.0                   0.0   
75%         0.000000  ...                       0.0                   0.0   
max         1.000000  ...                       0.0                   0.0   

       nose deformity  lump over jaw   sore in nose  hip weakness  \
count        246945.0       246945.0  246945.000000      246945.0   
mean              0.0            0.0       0.001385           0.0   
std               0.0            0.0       0.037189           0.0   
min               0.0            0.0       0.000000           0.0   
25%               0.0            0.0       0.000000           0.0   
50%               0.0            0.0       0.000000           0.0   
75%               0.0            0.0       0.000000           0.0   
max               0.0            0.0       1.000000           0.0   

       back swelling  ankle stiffness or tightness  ankle weakness  \
count       246945.0                      246945.0   246945.000000   
mean             0.0                           0.0        0.000073   
std              0.0                           0.0        0.008537   
min              0.0                           0.0        0.000000   
25%              0.0                           0.0        0.000000   
50%              0.0                           0.0        0.000000   
75%              0.0                           0.0        0.000000   
max              0.0                           0.0        1.000000   

       neck weakness  
count       246945.0  
mean             0.0  
std              

In [6]:
# Let's take a look at the number of diseases we have:
len(df["diseases"].unique())
# There are 773 of them, let's take a look some of them:
for disease in df_sample["diseases"].unique():
    print(f"number of {disease}: {df_sample["diseases"].value_counts()[disease]}")
# Some are numerous, others are few

number of cornea infection: 4
number of idiopathic painful menstruation: 6
number of cerebral palsy: 1
number of ovarian cyst: 2
number of skin polyp: 5
number of eustachian tube dysfunction (ear disorder): 5
number of cholecystitis: 5
number of prostatitis: 4
number of gastrointestinal hemorrhage: 4
number of gonorrhea: 1
number of tic (movement) disorder: 1
number of autism: 4
number of spondylitis: 1
number of viral warts: 2
number of aphthous ulcer: 2
number of esophagitis: 3
number of dry eye of unknown cause: 3
number of hypoglycemia: 6
number of prostate cancer: 1
number of schizophrenia: 7
number of acute bronchiolitis: 6
number of pulmonary congestion: 3
number of panic disorder: 4
number of callus: 4
number of paroxysmal ventricular tachycardia: 1
number of hyperkalemia: 3
number of interstitial lung disease: 1
number of flu: 4
number of diabetic ketoacidosis: 2
number of oral mucosal lesion: 4
number of pyelonephritis: 7
number of vitamin b12 deficiency: 1
number of drug abu

**Pre-processing**

Due to our poor skills in medicine, and to improve both efficiency and relevance, we will use an AI to store the diseases into larger, more comprehensive categories. We will ask it to do it just as we saw in class, with a "if ... in disease... yada yada".

In [7]:
def group_disease(disease):
    disease = disease.lower()
    
    # Maladies cardiovasculaires
    if any(x in disease for x in ["heart", "coronary", "hypertension", "cardiomyopathy", "aortic", "stroke", "ischemia", "atrial", "valve", "arrhythmia", "thrombosis", "embolism", "angina", "arterial", "vascular"]):
        return "cardiovascular"
    
    # Maladies respiratoires
    elif any(x in disease for x in ["asthma", "copd", "emphysema", "pulmonary", "bronchiolitis", "respiratory", "pneumonia", "bronchitis", "breath", "lung", "tracheal", "laryngeal", "pleural", "pneumothorax"]):
        return "respiratory"
    
    # Maladies métaboliques / endocriniennes
    elif any(x in disease for x in ["diabetes", "thyroid", "obesity", "metabolic", "hypoglycemia", "hyperglycemia", "parathyroid", "pituitary", "adrenal", "hormone", "cushing", "graves", "hashimoto"]):
        return "metabolic_endocrine"
    
    # Maladies neurologiques / mentales
    elif any(x in disease for x in ["depression", "anxiety", "parkinson", "dementia", "adhd", "migraine", "epilepsy", "neuropathy", "stroke", "disorder", "bipolar", "schizophrenia", "psychotic", "panic", "ptsd", "traumatic", "stress", "mental", "neurological", "brain", "alzheimer", "huntington", "als", "sclerosis", "palsy", "nerve"]):
        return "neurological_psych"
    
    # Maladies digestives / hépatiques
    elif any(x in disease for x in ["liver", "gastro", "pancreatitis", "esophagus", "stomach", "intestinal", "ulcer", "nash", "hepatitis", "crohn", "colitis", "ibs", "diverticular", "cirrhosis", "cholecyst", "bile", "bowel"]):
        return "digestive_hepatic"
    
    # Maladies rénales / urologiques
    elif any(x in disease for x in ["kidney", "renal", "urethral", "bladder", "prostate", "urine", "urinary", "urologic", "uric", "urea"]):
        return "renal_urological"
    
    # Maladies infectieuses
    elif any(x in disease for x in ["infection", "viral", "bacterial", "fungal", "dengue", "hiv", "malaria", "tuberculosis", "meningitis", "sepsis", "pneumonia", "hepatitis", "herpes", "gonorrhea", "chlamydia", "syphilis", "lyme", "fever", "cold", "flu", "wart", "lice", "scabies"]):
        return "infectious"
    
    # Maladies musculo-squelettiques
    elif any(x in disease for x in ["fracture", "injury", "arthritis", "osteoporosis", "spine", "knee", "shoulder", "ankle", "foot", "hip", "bone", "cartilage", "tendon", "ligament", "muscle", "sprain", "strain", "dislocation", "bursitis", "carpal tunnel"]):
        return "musculoskeletal"
    
    # Maladies dermatologiques
    elif any(x in disease for x in ["skin", "dermatitis", "eczema", "psoriasis", "rash", "acne", "wound", "burn", "ulcer", "sore", "wart", "fungal", "abscess", "cyst", "mole", "lesion", "blister", "impetigo"]):
        return "dermatological"
    
    # Maladies oncologiques
    elif any(x in disease for x in ["cancer", "tumor", "carcinoma", "lymphoma", "sarcoma", "leukemia", "malignant", "ependymoma", "meningioma", "adenoma", "polyp"]):
        return "oncology"
    
    # Maladies gynécologiques / obstétriques
    elif any(x in disease for x in ["pregnancy", "uterine", "ovarian", "endometriosis", "fibroids", "menopause", "vagina", "placenta", "cervix", "womb", "female", "gynecolog", "obstetric", "natal", "abortion", "ectopic", "preeclamp"]):
        return "gynecological_obstetric"
    
    # Maladies ophtalmologiques / auditives
    elif any(x in disease for x in ["eye", "glaucoma", "retina", "vision", "hearing", "tinnitus", "otitis", "ocular", "conjunctivitis", "cataract", "cornea", "optic", "visual", "deaf", "ear", "auditory"]):
        return "ophthalmology_audiology"
    
    # Maladies hématologiques / immunologiques
    elif any(x in disease for x in ["anemia", "bleeding", "coagulation", "immunodeficiency", "leukemia", "lymphoma", "sickle", "hemophil", "thrombocyt", "polycyth"]):
        return "hematological_immunological"
    
    # Autres / non classés
    else:
        return "other"

df["group diseases"] = df["diseases"].apply(group_disease)
#df_sample["group diseases"] = df["diseases"].apply(group_disease)

# In case here is the list of the above grouped categories:
#["cardiovascular", "respiratory", "metabolic_endocrine", "neurological_psych", "digestive_hepatic", "renal_urological", "infectious", "musculoskeletal", "dermatological", "oncology", "gynecological_obstetric", "ophthalmology_audiology", "hematological_immunological", "other"]

Now that we have grouped many of them, let's take some time to look at them again: 

In [13]:
df.head(10).style

for disease in df["group diseases"].unique():
    print(f"number of {disease}: {df["group diseases"].value_counts()[disease]}")

number of neurological_psych: 29611
number of oncology: 2162
number of other: 104838
number of musculoskeletal: 17141
number of dermatological: 16460
number of metabolic_endocrine: 2078
number of ophthalmology_audiology: 13379
number of gynecological_obstetric: 6289
number of digestive_hepatic: 12872
number of cardiovascular: 11142
number of respiratory: 8220
number of hematological_immunological: 2635
number of infectious: 14021
number of renal_urological: 6097


Should we remove all the other category? I think so, as they are not relevant to compute our tumor, but we will see that.

Now, lets turn to our features: the symptoms. They may be too numerous. The IA mention doing a variance threshold method, I look at the documentation on the Internet and try to apply it with the help of the AI.
For the AI, a threshold of 0.01 or 0.05 seem relevant. In any case, the goal is to remove features that are too constant between diseases to be relevant; they do not allow to differentiate between disease

In [14]:
from sklearn.feature_selection import VarianceThreshold
# First, let's identify which columns are symptom features (not metadata)
# Assuming 'diseases' and 'group diseases' are the target/metadata columns

symptom_columns = [col for col in df.columns if col not in ['diseases', 'group diseases']]
X = df[symptom_columns]

print(f"Initial number of symptoms: {len(symptom_columns)}")
print(f"Dataset shape: {X.shape}")

# Apply Variance Threshold
# For binary features (0/1), threshold is based on Var(X) = p(1-p)
# threshold=0.01 removes features that appear in less than ~1% or more than ~99% of cases
selector = VarianceThreshold(threshold=0.01)
X_filtered = selector.fit_transform(X)

# Get the names of selected features
selected_features = X.columns[selector.get_support()].tolist()
removed_features = X.columns[~selector.get_support()].tolist()

print(f"\nAfter variance threshold (0.01):")
print(f"Number of symptoms remaining: {len(selected_features)}")
print(f"Number of symptoms removed: {len(removed_features)}")
print(f"New shape: {X_filtered.shape}")

# Show some removed features (the ones with very low variance)
print(f"\nSample of removed symptoms (low variance):")
print(removed_features[:10])


Initial number of symptoms: 377
Dataset shape: (246945, 377)

After variance threshold (0.01):
Number of symptoms remaining: 146
Number of symptoms removed: 231
New shape: (246945, 146)

Sample of removed symptoms (low variance):
['breathing fast', 'throat swelling', 'lump in throat', 'throat feels tight', 'groin mass', 'emotional symptoms', 'elbow weakness', 'back weakness', 'pus in sputum', 'symptoms of the scrotum and testes']

After variance threshold (0.01):
Number of symptoms remaining: 146
Number of symptoms removed: 231
New shape: (246945, 146)

Sample of removed symptoms (low variance):
['breathing fast', 'throat swelling', 'lump in throat', 'throat feels tight', 'groin mass', 'emotional symptoms', 'elbow weakness', 'back weakness', 'pus in sputum', 'symptoms of the scrotum and testes']


In [17]:
# Create a new dataframe with only selected symptoms + target/metadata columns
df_cleaned = df[['group diseases'] + selected_features].copy()

print(f"Cleaned dataset shape: {df_cleaned.shape}")
print(f"\nCleaned dataset columns: {df_cleaned.columns.tolist()[:10]}... (showing first 10)")

# You can adjust the threshold if needed:
# threshold=0.001 -> removes very rare symptoms (stricter)
# threshold=0.05  -> removes more symptoms (less strict)
# threshold=0.0   -> keeps all features with any variance


Cleaned dataset shape: (246945, 147)

Cleaned dataset columns: ['group diseases', 'anxiety and nervousness', 'depression', 'shortness of breath', 'depressive or psychotic symptoms', 'sharp chest pain', 'dizziness', 'insomnia', 'abnormal involuntary movements', 'chest tightness']... (showing first 10)


In [18]:
df_cleaned

group diseases  anxiety and nervousness  depression  \
0       neurological_psych                        1           0   
1       neurological_psych                        0           0   
2       neurological_psych                        1           1   
3       neurological_psych                        1           0   
4       neurological_psych                        1           1   
...                    ...                      ...         ...   
246940      dermatological                        0           0   
246941      dermatological                        0           0   
246942      dermatological                        0           0   
246943      dermatological                        0           0   
246944      dermatological                        0           0   

        shortness of breath  depressive or psychotic symptoms  \
0                         1                                 1   
1                         1                                 1   
2                         1                                 1   
3                         0                                 1   
4                         0                                 0   
...                     ...                               ...   
246940                    0                                 0   
246941                    0                                 0   
246942                    0                                 0   
246943                    0                                 0   
246944                    0                                 0   

        sharp chest pain  dizziness  insomnia  abnormal involuntary movements  \
0                      0          0         0                               0   
1                      0          1         1                               0   
2                      0          1         1                               0   
3                      0          1         1                               1   
4                      0          0         1                               1   
...                  ...        ...       ...                             ...   
246940                 0          0         0                               0   
246941                 0          0         0                               0   
246942                 0          0         0                               0   
246943                 0          0         0                               0   
246944                 0          0         0                               0   

        chest tightness  ...  sweating  swollen eye  symptoms of bladder  \
0                     1  ...         0            0                    0   
1                     0  ...         0            0                    0   
2                     0  ...         0            0                    0   
3                     0  ...         0            0                    0   
4                     1  ...         0            0                    0   
...                 ...  ...       ...          ...                  ...   
246940                0  ...         0            0                    0   
246941                0  ...         0            0                    0   
246942                0  ...         0            0                    0   
246943                0  ...         0            0                    0   
246944                0  ...         0            0                    0   

        itching of skin  skin dryness, peeling, scaliness, or roughness  \
0                     0                                               0   
1                     0                                               0   
2                     0                                               0   
3                     0                                               0   
4                     0                                               0   
...                 ...                                             ...   
246940         

Now we are going to train the model, just as in the colortree exercise, to predict the disease from the syptoms. 

In [ ]:
# Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

# Instanciate needed parameters and set the random seed to be reproductible
test_ratio = 0.2
seed_n = 42
np.random.seed(seed_n)

# Define X and Y
X = df_cleaned[[col for col in df_cleaned.columns if col not in ['group diseases']]]
Y = df_cleaned["group diseases"]

# Split the dataset into training and testing sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=test_ratio, random_state=seed_n)

SyntaxError: invalid syntax (412650589.py, line 14)